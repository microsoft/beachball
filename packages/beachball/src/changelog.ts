import { PackageInfo } from './bump';
import path from 'path';
import fs from 'fs';
import { generateTag } from './tag';
import { ChangeInfo, ChangeSet } from './ChangeInfo';

interface ChangelogEntry {
  comment: string;
  author: string;
  commit: string;
}

interface PackageChangelog {
  name: string;
  date: Date;
  version: string;
  comments: {
    prerelease?: ChangelogEntry[];
    patch?: ChangelogEntry[];
    minor?: ChangelogEntry[];
    major?: ChangelogEntry[];
    none?: ChangelogEntry[];
  };
}

interface ChangelogJsonEntry {
  date: string;
  version: string;
  tag: string;
  comments: {
    prerelease?: ChangelogEntry[];
    patch?: ChangelogEntry[];
    minor?: ChangelogEntry[];
    major?: ChangelogEntry[];
    none?: ChangelogEntry[];
  };
}

interface ChangelogJson {
  name: string;
  entries: ChangelogJsonEntry[];
}

export function getPackageChangelogs(changeSet: ChangeSet, packageInfos: { [pkg: string]: PackageInfo }) {
  const changelogs: { [pkgName: string]: PackageChangelog } = {};

  for (let [_, change] of changeSet) {
    const { packageName } = change;

    changelogs[packageName] = changelogs[packageName] || {
      name: packageName,
      version: packageInfos[packageName].version,
      date: new Date(),
    };

    changelogs[packageName].comments = changelogs[packageName].comments || {};
    changelogs[packageName].comments[change.type] = changelogs[packageName].comments[change.type] || [];
    changelogs[packageName].comments[change.type]!.push({
      comment: change.comment,
      author: change.email,
      commit: change.commit,
    });
  }

  return changelogs;
}

export function writeChangelog(changeSet: ChangeSet, packageInfos: { [pkg: string]: PackageInfo }) {
  const changelogs = getPackageChangelogs(changeSet, packageInfos);

  Object.keys(changelogs).forEach(pkg => {
    const packagePath = path.dirname(packageInfos[pkg].packageJsonPath);

    if (
      changelogs[pkg].comments.major ||
      changelogs[pkg].comments.minor ||
      changelogs[pkg].comments.patch ||
      changelogs[pkg].comments.prerelease
    ) {
      const changelogFile = path.join(packagePath, 'CHANGELOG.md');
      const previousContent = fs.existsSync(changelogFile) ? fs.readFileSync(changelogFile).toString() : '';
      const nextContent = renderChangelog(previousContent, changelogs[pkg]);
      fs.writeFileSync(changelogFile, nextContent);
    }

    try {
      const changelogJsonFile = path.join(packagePath, 'CHANGELOG.json');
      const previousJson = fs.existsSync(changelogJsonFile)
        ? JSON.parse(fs.readFileSync(changelogJsonFile).toString())
        : { entries: [] };
      const nextJson = renderJsonChangelog(previousJson, changelogs[pkg]);
      fs.writeFileSync(changelogJsonFile, JSON.stringify(nextJson, null, 2));
    } catch (e) {
      console.warn('The CHANGELOG.json file is invalid, skipping writing to it', e);
    }
  });
}

function renderJsonChangelog(previous: ChangelogJson, changelog: PackageChangelog) {
  const result: ChangelogJson = {
    name: changelog.name,
    entries: [...previous.entries] || [],
  };

  const newEntry: ChangelogJsonEntry = {
    date: changelog.date.toUTCString(),
    tag: generateTag(changelog.name, changelog.version),
    version: changelog.version,
    comments: changelog.comments,
  };

  result.entries.unshift(newEntry);

  return result;
}

function renderChangelog(previous: string, changelog: PackageChangelog) {
  const previousLogEntries = previous ? '\n' + previous.substring(previous.indexOf('##')) : '';

  return (
    `# Change Log - ${changelog.name}\n\n` +
    `This log was last generated on ${changelog.date.toUTCString()} and should not be manually modified.\n` +
    renderPackageChangelog(changelog) +
    previousLogEntries
  );
}

function renderPackageChangelog(changelog: PackageChangelog) {
  return (
    `\n## ${changelog.version}\n` +
    `${changelog.date.toUTCString()}\n` +
    (changelog.comments.major
      ? '\n### Major\n\n' + changelog.comments.major.map(change => `- ${change.comment} (${change.author})`).join('\n')
      : '') +
    (changelog.comments.minor
      ? '\n### Minor changes\n\n' +
        changelog.comments.minor.map(change => `- ${change.comment} (${change.author})`).join('\n')
      : '') +
    (changelog.comments.patch
      ? '\n### Patches\n\n' +
        changelog.comments.patch.map(change => `- ${change.comment} (${change.author})`).join('\n')
      : '') +
    (changelog.comments.prerelease
      ? '\n### Changes\n\n' +
        changelog.comments.prerelease.map(change => `- ${change.comment} (${change.author})`).join('\n')
      : '')
  );
}
