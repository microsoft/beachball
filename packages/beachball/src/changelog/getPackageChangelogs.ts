import { ChangeSet } from '../types/ChangeInfo';
import { PackageInfo } from '../types/PackageInfo';
import { PackageChangelog } from '../types/ChangeLog';
export function getPackageChangelogs(changeSet: ChangeSet, packageInfos: {
  [pkg: string]: PackageInfo;
}) {
  const changelogs: {
    [pkgName: string]: PackageChangelog;
  } = {};
  for (let [_, change] of changeSet) {
    const { packageName } = change;
    changelogs[packageName] = changelogs[packageName] || {
      name: packageName,
      version: packageInfos[packageName].version,
      date: new Date(),
    };
    changelogs[packageName].comments = changelogs[packageName].comments || {};
    changelogs[packageName].comments[change.type] = changelogs[packageName].comments[change.type] || [];
    changelogs[packageName].comments[change.type]!.push({
      comment: change.comment,
      author: change.email,
      commit: change.commit,
    });
  }
  return changelogs;
}
