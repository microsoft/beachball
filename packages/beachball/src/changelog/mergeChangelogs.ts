import _ from 'lodash';
import { PackageChangelog } from '../types/ChangeLog';
import { PackageInfo } from '../types/PackageInfo';

/**
 * Merge multiple PackageChangelog into one.
 * `name`, `date` and `version` will be using the values from master changelog. `comments` are merged.
 */
export function mergeChangelogs(
  changelogs: PackageChangelog[],
  masterPackage: PackageInfo
): PackageChangelog | undefined {
  if (changelogs.length < 1 || !masterPackage) {
    return undefined;
  }

  const result: PackageChangelog = {
    name: masterPackage.name,
    version: masterPackage.version,
    date: new Date(),
    comments: {},
  };

  changelogs.forEach(changelog => {
    Object.keys(changelog.comments).forEach(changeType => {
      if (changelog.comments[changeType]) {
        result.comments[changeType] = (result.comments[changeType] || []).concat(changelog.comments[changeType]);
      }
    });
  });

  return result;
}
