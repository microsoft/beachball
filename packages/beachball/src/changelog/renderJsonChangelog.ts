import { generateTag } from '../tag';
import { PackageChangelog, ChangelogJson, ChangelogJsonEntry } from '../types/ChangeLog';

export function renderJsonChangelog(changelog: PackageChangelog, previousChangelog: ChangelogJson | undefined) {
  const result: ChangelogJson = {
    name: changelog.name,
    entries: previousChangelog?.entries ? [...previousChangelog.entries] : [],
  };
  const newEntry: ChangelogJsonEntry = {
    date: changelog.date.toUTCString(),
    tag: generateTag(changelog.name, changelog.version),
    version: changelog.version,
    comments: changelog.comments,
  };
  result.entries.unshift(newEntry);
  return result;
}
