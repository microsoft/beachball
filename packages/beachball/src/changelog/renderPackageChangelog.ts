import { PackageChangelog, ChangelogEntry } from '../types/ChangeLog';
import _ from 'lodash';

export function renderPackageChangelog(changelog: PackageChangelog, isGroupedChangelog: boolean = false) {
  return (
    `\n## ${changelog.version}\n` +
    `${changelog.date.toUTCString()}\n` +
    (changelog.comments.major
      ? '\n### Major\n\n' + renderChangelogEntries(changelog.comments.major, isGroupedChangelog)
      : '') +
    (changelog.comments.minor
      ? '\n### Minor changes\n\n' + renderChangelogEntries(changelog.comments.minor, isGroupedChangelog)
      : '') +
    (changelog.comments.patch
      ? '\n### Patches\n\n' + renderChangelogEntries(changelog.comments.patch, isGroupedChangelog)
      : '') +
    (changelog.comments.prerelease
      ? '\n### Changes\n\n' + renderChangelogEntries(changelog.comments.prerelease, isGroupedChangelog)
      : '')
  );
}

function renderChangelogEntries(entries: ChangelogEntry[], includePackageInfo: boolean = false): string {
  if (includePackageInfo) {
    const entriesMap = _.groupBy(entries, entry => entry.package);

    let result = '';
    Object.keys(entriesMap).forEach(pkgName => {
      const entries = entriesMap[pkgName];
      result += `- \`${pkgName}\`\n`;
      entries.forEach(entry => {
        result += `  - ${entry.comment} (${entry.author})\n`;
      });
    });

    return result;
  }

  return entries
    .map(entry => {
      return `- ${entry.comment} (${entry.author})`;
    })
    .join('\n');
}
