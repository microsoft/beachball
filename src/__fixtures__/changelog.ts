import fs from 'fs-extra';
import path from 'path';
import _ from 'lodash';
import { SortedChangeTypes } from '../changefile/changeTypes';
import { ChangelogJson } from '../types/ChangeLog';

/** Read the CHANGELOG.md under the given package path, sanitizing any dates for snapshots */
export function readChangelogMd(packagePath: string): string {
  const changelogFile = path.join(packagePath, 'CHANGELOG.md');
  const text = fs.readFileSync(changelogFile, { encoding: 'utf-8' });
  return text.replace(/\w\w\w, \d\d \w\w\w [\d :]+?GMT/gm, '(date)');
}

/** Read the CHANGELOG.json under the given package path */
export function readChangelogJson(packagePath: string, cleanForSnapshot: boolean = false): ChangelogJson {
  const changelogJsonFile = path.join(packagePath, 'CHANGELOG.json');
  const json = fs.readJSONSync(changelogJsonFile, { encoding: 'utf-8' });
  return cleanForSnapshot ? cleanChangelogJson(json) : json;
}

/**
 * Clean changelog json for a snapshot: replace dates and SHAs with placeholders.
 * Note: this clones the changelog object rather than modifying the original.
 */
export function cleanChangelogJson(changelog: ChangelogJson): ChangelogJson {
  changelog = _.cloneDeep(changelog);
  // for a better snapshot, make the fake commit match if the real commit did
  const fakeCommits: { [commit: string]: string } = {};
  let fakeHashNum = 0;

  for (const entry of changelog.entries) {
    entry.date = '(date)';
    for (const changeType of SortedChangeTypes) {
      if (entry.comments[changeType]) {
        for (const comment of entry.comments[changeType]!) {
          if (!fakeCommits[comment.commit]) {
            fakeCommits[comment.commit] = `(sha1-${fakeHashNum++})`;
          }
          comment.commit = fakeCommits[comment.commit];
        }
      }
    }
  }
  return changelog;
}
