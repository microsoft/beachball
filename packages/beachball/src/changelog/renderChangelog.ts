import { PackageChangelog } from '../types/ChangeLog';
import { renderPackageChangelog, defaultRenderers } from './renderPackageChangelog';
import { ChangelogOptions, PackageChangelogRenderInfo } from '../types/ChangelogOptions';

export interface MarkdownChangelogRenderOptions extends Omit<PackageChangelogRenderInfo, 'renderers'> {
  previousContent: string;
  changelogOptions: ChangelogOptions;
}

export const markerComment = '<!-- Start content -->';

export async function renderChangelog(renderOptions: MarkdownChangelogRenderOptions): Promise<string> {
  const {
    previousJson,
    previousContent = '',
    newVersionChangelog,
    isGrouped,
    changelogOptions: { renderPackageChangelog: customRenderPackageChangelog, customRenderers },
  } = renderOptions;

  let previousLogEntries: string;
  if (previousContent.includes(markerComment)) {
    // Preferably determine where the previous entries start based on a special comment
    previousLogEntries = previousContent.split(markerComment, 2)[1].trim();
  } else {
    // Otherwise look for an h2 (used as version header with default renderer).
    // If that's not present, preserve the previous content as-is.
    const h2Match = previousContent.match(/^## /m);
    previousLogEntries = h2Match ? previousContent.substring(h2Match.index!) : previousContent;
  }

  try {
    if (customRenderPackageChangelog || customRenderers) {
      console.log('Using custom renderer for package version changelog.');
    }

    const renderInfo: PackageChangelogRenderInfo = {
      previousJson,
      newVersionChangelog,
      isGrouped,
      renderers: {
        ...defaultRenderers,
        ...customRenderers,
      },
    };

    return (
      [
        renderChangelogHeader(newVersionChangelog),
        markerComment,
        await (customRenderPackageChangelog || renderPackageChangelog)(renderInfo),
        previousLogEntries,
      ]
        .join('\n\n')
        .trim() + '\n'
    );
  } catch (err) {
    console.log('Error occurred rendering package version changelog:', err);
    return '';
  }
}

function renderChangelogHeader(changelog: PackageChangelog): string {
  return (
    `# Change Log - ${changelog.name}\n\n` +
    `This log was last generated on ${changelog.date.toUTCString()} and should not be manually modified.`
  );
}
