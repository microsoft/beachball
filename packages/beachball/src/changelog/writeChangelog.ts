import path from 'path';
import fs from 'fs-extra';
import _ from 'lodash';
import { ChangeSet } from '../types/ChangeInfo';
import { PackageInfo } from '../types/PackageInfo';
import { getPackageChangelogs } from './getPackageChangelogs';
import { renderChangelog } from './renderChangelog';
import { renderJsonChangelog } from './renderJsonChangelog';
import { BeachballOptions } from '../types/BeachballOptions';
import { isPathIncluded } from '../monorepo/utils';
import { PackageChangelog, ChangelogJson } from '../types/ChangeLog';
import { mergeChangelogs } from './mergeChangelogs';

export async function writeChangelog(
  options: BeachballOptions,
  changeSet: ChangeSet,
  packageInfos: {
    [pkg: string]: PackageInfo;
  }
): Promise<void> {
  const groupedChangelogPaths = await writeGroupedChangelog(options, changeSet, packageInfos);
  const groupedChangelogPathSet = new Set(groupedChangelogPaths);

  const changelogs = getPackageChangelogs(changeSet, packageInfos);
  // Use a standard for loop here to prevent potentially firing off multiple network requests at once
  // (in case any custom renderers have network requests)
  for (const pkg of Object.keys(changelogs)) {
    const packagePath = path.dirname(packageInfos[pkg].packageJsonPath);
    if (groupedChangelogPathSet?.has(packagePath)) {
      console.log(`Changelog for ${pkg} has been written as a group here: ${packagePath}`);
    } else {
      await writeChangelogFiles(options, changelogs[pkg], packagePath, false);
    }
  }
}

async function writeGroupedChangelog(
  options: BeachballOptions,
  changeSet: ChangeSet,
  packageInfos: {
    [pkg: string]: PackageInfo;
  }
): Promise<string[]> {
  if (!options.changelog) {
    return [];
  }

  const { groups: changelogGroups } = options.changelog;
  if (!changelogGroups || changelogGroups.length < 1) {
    return [];
  }

  const changelogs = getPackageChangelogs(changeSet, packageInfos);
  const groupedChangelogs: {
    [path: string]: { changelogs: PackageChangelog[]; masterPackage: PackageInfo };
  } = {};

  for (const pkg in changelogs) {
    const packagePath = path.dirname(packageInfos[pkg].packageJsonPath);
    const relativePath = path.relative(options.path, packagePath);
    for (const group of changelogGroups) {
      const { changelogPath, masterPackageName } = group;
      const masterPackage = packageInfos[masterPackageName];
      if (!masterPackage) {
        console.warn(`master package ${masterPackageName} does not exist.`);
        continue;
      }
      if (!fs.existsSync(changelogPath)) {
        console.warn(`changelog path ${changelogPath} does not exist.`);
        continue;
      }

      const isInGroup = isPathIncluded(relativePath, group.include, group.exclude);
      if (isInGroup) {
        if (!groupedChangelogs[changelogPath]) {
          groupedChangelogs[changelogPath] = {
            changelogs: [],
            masterPackage,
          };
        }

        groupedChangelogs[changelogPath].changelogs.push(changelogs[pkg]);
      }
    }
  }

  const changelogAbsolutePaths: string[] = [];
  for (const changelogPath in groupedChangelogs) {
    const { masterPackage, changelogs } = groupedChangelogs[changelogPath];
    const groupedChangelog = mergeChangelogs(changelogs, masterPackage);
    if (groupedChangelog) {
      await writeChangelogFiles(options, groupedChangelog, changelogPath, true);
      changelogAbsolutePaths.push(path.resolve(changelogPath));
    }
  }

  return changelogAbsolutePaths;
}

async function writeChangelogFiles(
  options: BeachballOptions,
  newVersionChangelog: PackageChangelog,
  changelogPath: string,
  isGrouped: boolean
): Promise<void> {
  let previousJson: ChangelogJson | undefined;

  // Update CHANGELOG.json
  const changelogJsonFile = path.join(changelogPath, 'CHANGELOG.json');
  try {
    previousJson = fs.existsSync(changelogJsonFile) ? fs.readJSONSync(changelogJsonFile) : undefined;
  } catch (e) {
    console.warn('CHANGELOG.json is invalid:', e);
  }
  try {
    const nextJson = renderJsonChangelog(newVersionChangelog, previousJson);
    fs.writeJSONSync(changelogJsonFile, nextJson, { spaces: 2 });
  } catch (e) {
    console.warn('Problem writing to CHANGELOG.json:', e);
  }

  // Update CHANGELOG.md
  if (
    newVersionChangelog.comments.major ||
    newVersionChangelog.comments.minor ||
    newVersionChangelog.comments.patch ||
    newVersionChangelog.comments.prerelease
  ) {
    const changelogFile = path.join(changelogPath, 'CHANGELOG.md');
    const previousContent = fs.existsSync(changelogFile) ? fs.readFileSync(changelogFile).toString() : '';

    const newChangelog = await renderChangelog({
      previousJson,
      previousContent,
      newVersionChangelog,
      isGrouped,
      changelogOptions: options.changelog || {},
    });

    fs.writeFileSync(changelogFile, newChangelog);
  }
}
