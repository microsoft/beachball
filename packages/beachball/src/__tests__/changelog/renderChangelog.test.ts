import { MarkdownChangelogRenderOptions, renderChangelog } from '../../changelog/renderChangelog';

const previousContent = `# Change Log - foo

This log was last generated on Wed, 21 Aug 2019 21:20:40 GMT and should not be manually modified.

## 1.2.0

(content here)
`;

describe('renderChangelog', () => {
  function getOptions(): MarkdownChangelogRenderOptions {
    return {
      isGrouped: false,
      newVersionChangelog: {
        date: new Date('Thu Aug 22 2019 14:20:40 GMT-0700 (Pacific Daylight Time)'),
        name: 'foo',
        tag: 'foo_v1.2.3',
        version: '1.2.3',
        comments: {
          major: [],
          minor: [
            { comment: 'Awesome change', author: 'user1@example.com', commit: 'sha1', package: 'foo' },
            { comment: 'Boring change', author: 'user2@example.com', commit: 'sha2', package: 'foo' },
          ],
          patch: [
            { comment: 'Fix', author: 'user1@example.com', commit: 'sha3', package: 'foo' },
            { comment: 'stuff', author: 'user2@example.com', commit: 'sha4', package: 'foo' },
          ],
        },
      },
      previousContent,
      previousJson: {} as any,
      changelogOptions: {},
    };
  }

  it('handles no previous content', () => {
    const options = getOptions();
    options.previousContent = '';
    expect(renderChangelog(options)).toMatchSnapshot();
  });

  it('merges with previous content', () => {
    const options = getOptions();
    const result = renderChangelog(options);
    expect(result).toContain('last generated on Thu, 22 Aug 2019 21:20:40 GMT'); // uses new date
    expect(result).toMatchSnapshot();
  });

  it('merges default and custom renderers', () => {
    const options = getOptions();
    options.changelogOptions = {
      customRenderers: {
        renderHeader: renderInfo => {
          return [
            `## ${renderInfo.newVersionChangelog.version}`,
            renderInfo.newVersionChangelog.date.toUTCString(),
            `[Compare changes](http://real-github-compare-link)`,
          ].join('\n');
        },
        renderEntry: (entry, renderInfo) => `- ${entry.comment} (${entry.author}, PR #123)`,
      },
    };

    const result = renderChangelog(options);
    expect(result).toContain('Compare changes');
    expect(result).toContain('PR #123');
    expect(result).toMatchSnapshot();
  });

  it('uses full custom renderer', () => {
    const options = getOptions();
    options.changelogOptions = {
      renderPackageChangelog: renderInfo => `## ${renderInfo.newVersionChangelog.version}\n\nno notes for you`,
    };

    const result = renderChangelog(options);
    expect(result).toContain('# Change Log - foo'); // still includes header
    expect(result).toContain('no notes for you'); // uses custom version body
    expect(result).toContain('content here'); // includes previous content
    expect(result).toMatchSnapshot();
  });
});
