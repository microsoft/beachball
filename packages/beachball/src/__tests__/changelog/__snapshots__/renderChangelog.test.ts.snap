// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderChangelog handles no previous content 1`] = `
"# Change Log - foo

This log was last generated on Thu, 22 Aug 2019 21:20:40 GMT and should not be manually modified.

<!-- Start content -->

## 1.2.3

Thu, 22 Aug 2019 21:20:40 GMT

### Minor changes

- Awesome change (user1@example.com)
- Boring change (user2@example.com)

### Patches

- Fix (user1@example.com)
- stuff (user2@example.com)
"
`;

exports[`renderChangelog keeps previous content if no marker or h2 is found 1`] = `
"# Change Log - foo

This log was last generated on Thu, 22 Aug 2019 21:20:40 GMT and should not be manually modified.

<!-- Start content -->

## 1.2.3

Thu, 22 Aug 2019 21:20:40 GMT

### Minor changes

- Awesome change (user1@example.com)
- Boring change (user2@example.com)

### Patches

- Fix (user1@example.com)
- stuff (user2@example.com)

# Change Log - foo

This log was last generated on Wed, 21 Aug 2019 21:20:40 GMT and should not be manually modified.
"
`;

exports[`renderChangelog merges default and custom renderers 1`] = `
"# Change Log - foo

This log was last generated on Thu, 22 Aug 2019 21:20:40 GMT and should not be manually modified.

<!-- Start content -->

## 1.2.3
Thu, 22 Aug 2019 21:20:40 GMT
[Compare changes](http://real-github-compare-link)

### Minor changes

- Awesome change (user1@example.com, PR #123)
- Boring change (user2@example.com, PR #123)

### Patches

- Fix (user1@example.com, PR #123)
- stuff (user2@example.com, PR #123)

## 1.2.0

(content here)
"
`;

exports[`renderChangelog merges with previous content using h2 1`] = `
"# Change Log - foo

This log was last generated on Thu, 22 Aug 2019 21:20:40 GMT and should not be manually modified.

<!-- Start content -->

## 1.2.3

Thu, 22 Aug 2019 21:20:40 GMT

### Minor changes

- Awesome change (user1@example.com)
- Boring change (user2@example.com)

### Patches

- Fix (user1@example.com)
- stuff (user2@example.com)

## 1.2.0

(content here)
"
`;

exports[`renderChangelog merges with previous content using marker 1`] = `
"# Change Log - foo

This log was last generated on Thu, 22 Aug 2019 21:20:40 GMT and should not be manually modified.

<!-- Start content -->

## 1.2.3

Thu, 22 Aug 2019 21:20:40 GMT

### Minor changes

- Awesome change (user1@example.com)
- Boring change (user2@example.com)

### Patches

- Fix (user1@example.com)
- stuff (user2@example.com)

## 1.2.0

(content here)
"
`;

exports[`renderChangelog uses full custom renderer 1`] = `
"# Change Log - foo

This log was last generated on Thu, 22 Aug 2019 21:20:40 GMT and should not be manually modified.

<!-- Start content -->

## 1.2.3

no notes for you

## 1.2.0

(content here)
"
`;
